csharp/src/Drivers/BigQuery/BigQueryConnection.cs

```diff
@@ -0,0 +1,954 @@
+                nullBitmapBuffer.Append(true);
+                length++;
+            }
+
+            List<IArrowArray> dataArrays = new List<IArrowArray>
+            {
+                constraintFkCatalogBuilder.Build(),
+                constraintFkDbSchemaBuilder.Build(),
+                constraintFkTableBuilder.Build(),
+                constraintFkColumnNameBuilder.Build()
+            };
+
+            return new StructArray(
+                new StructType(StandardSchemas.UsageSchema),
+                length,
+                dataArrays,
+                nullBitmapBuffer.Build());
+        }
+
+        private string PatternToRegEx(string pattern)
+        {
+            if (pattern == null)
+                return ".*";
+
+            StringBuilder builder = new StringBuilder("(?i)^");
+            string convertedPattern = pattern.Replace("_", ".").Replace("%", ".*");
+            builder.Append(convertedPattern);
+            builder.Append("$");
+
+            return builder.ToString();
+        }
+
+        private string ToTypeName(string type)
+        {
+            int index = Math.Min(type.IndexOf("("), type.IndexOf("<"));
+            string dataType = index == -1 ? type : type.Substring(0, index);
+            return dataType;
+        }
+
+        private XdbcDataType ToXdbcDataType(string type)
+        {
+            switch (type)
+            {
+                case "INTEGER" or "INT64":
+                    return XdbcDataType.XdbcDataType_XDBC_INTEGER;
+                case "FLOAT" or "FLOAT64":
+                    return XdbcDataType.XdbcDataType_XDBC_FLOAT;
+                case "BOOL" or "BOOLEAN":
+                    return XdbcDataType.XdbcDataType_XDBC_BIT;
+                case "STRING":
+                    return XdbcDataType.XdbcDataType_XDBC_VARCHAR;
+                case "BYTES":
+                    return XdbcDataType.XdbcDataType_XDBC_BINARY;
+                case "DATETIME":
+                    return XdbcDataType.XdbcDataType_XDBC_DATETIME;
+                case "TIMESTAMP":
+                    return XdbcDataType.XdbcDataType_XDBC_TIMESTAMP;
+                case "TIME":
+                    return XdbcDataType.XdbcDataType_XDBC_TIME;
+                case "DATE":
+                    return XdbcDataType.XdbcDataType_XDBC_DATE;
+                case "RECORD" or "STRUCT":
+                    return XdbcDataType.XdbcDataType_XDBC_VARBINARY;
+                case "NUMERIC" or "DECIMAL" or "BIGNUMERIC" or "BIGDECIMAL":
+                    return XdbcDataType.XdbcDataType_XDBC_NUMERIC;
+
+                default:
+                    return XdbcDataType.XdbcDataType_XDBC_UNKNOWN_TYPE;
+            }
+        }
+
+        public override Schema GetTableSchema(string catalog, string dbSchema, string tableName)
+        {
+            string query = string.Format("SELECT * FROM `{0}`.`{1}`.INFORMATION_SCHEMA.COLUMNS WHERE table_name = '{2}'",
+                catalog, dbSchema, tableName);
```

Comment:
Out of curiosity (and not because it should block anything) have you investigated whether the schema hierarchy can be combined into a single query? BigQuery has pretty decent support for structured columns, so it ought to be possible to format things together in the query and reduce round trips. (My assumption is that this would be faster than non-concurrent round trips; that would need to be tested, of course.)

Reply:
@jiachen9393?